Properly document keepalive and other tcp options (#60216)
elastic/elasticsearch: docs/reference/modules/transport.asciidoc

to correctly configure TCP keep-alives instead of using this feature, because
TCP keep-alives apply to all kinds of long-lived connections and not just to
transport connections.

`transport.tcp.no_delay`::
Enable or disable the https://en.wikipedia.org/wiki/Nagle%27s_algorithm[TCP no delay]
setting. Defaults to `network.tcp.no_delay`.

`transport.tcp.keep_alive`::
Configures the `SO_KEEPALIVE` option for this socket, which
determines whether it sends TCP keepalive probes.
Defaults to `network.tcp.keep_alive`.

`transport.tcp.keep_idle`::
Configures the `TCP_KEEPIDLE` option for this socket, which
determines the time in seconds that a connection must be idle before
starting to send TCP keepalive probes. Defaults to `network.tcp.keep_idle`, which
is the system default. May not be greater than 300. In case where the system default
is higher than 300, the value is automatically lowered to 300. Only applicable on
Linux and macOS, and requires Java 11 or newer.

`transport.tcp.keep_interval`::
Configures the `TCP_KEEPINTVL` option for this socket,
which determines the time in seconds between sending TCP keepalive probes.

> Defaults to `network.tcp.keep_interval`, which is the system default.

Reviewer: This is awkward to explain; my best idea so far is this:

```suggestion
Defaults to `network.tcp.keep_interval` if set, or the system default otherwise.
```

Maybe the docs folks can think of a better way to explain this?

May not be greater than 300. In case where the system default is higher than 300,
the value is automatically lowered to 300. Only applicable on Linux and macOS,
and requires Java 11 or newer.

`transport.tcp.keep_count`::
Configures the `TCP_KEEPCNT` option for this socket, which
determines the number of unacknowledged TCP keepalive probes that may be
sent on a connection before it is dropped. Defaults to `network.tcp.keep_count`

> which means to use the system default. Only applicable on Linux and macOS, and

Reviewer: ```suggestion
if set, or the system default otherwise. Only applicable on Linux and macOS, and
```

[[transport-settings]]
==== Transport settings

The following settings can be configured for the internal transport that
communicates over TCP. These settings also use the common
<<modules-network,network settings>>.

`transport.port`::

> A bind port range. Defaults to `9300-9400`.

Reviewer 2: Thanks for fixing the formatting here!

`transport.connect_timeout`::
The connect timeout for initiating a new connection (in
time setting format). Defaults to `30s`.

`transport.compress`::
Set to `true` to enable compression (`DEFLATE`) between
all nodes. Defaults to `false`.

`transport.ping_schedule`::
Schedule a regular application-level ping message
to ensure that transport connections between nodes are kept alive. Defaults to
`5s` in the transport client and `-1` (disabled) elsewhere. It is preferable
to correctly configure TCP keep-alives instead of using this feature, because
TCP keep-alives apply to all kinds of long-lived connections and not just to
transport connections.

`transport.tcp.no_delay`::
Enable or disable the https://en.wikipedia.org/wiki/Nagle%27s_algorithm[TCP no delay]
setting. Defaults to `network.tcp.no_delay`.

`transport.tcp.keep_alive`::
Configures the `SO_KEEPALIVE` option for this socket, which
determines whether it sends TCP keepalive probes.
Defaults to `network.tcp.keep_alive`.

`transport.tcp.keep_idle`::
Configures the `TCP_KEEPIDLE` option for this socket, which
determines the time in seconds that a connection must be idle before

> starting to send TCP keepalive probes. Defaults to `network.tcp.keep_idle`, which
> Defaults to `network.tcp.keep_idle` if set, or the system default otherwise.

Reviewer 2: ```suggestion
starting to send TCP keepalive probes. Defaults to `network.tcp.keep_idle` if set,
or the system default otherwise.
```

> May not be greater than 300. In case where the system default
> is higher than 300, the value is automatically lowered to 300. Only applicable on

Reviewer 2: ```suggestion
This value cannot exceed `300` seconds. In cases where the system default
is higher than `300`, the value is automatically lowered to `300`. Only applicable on
```

Linux and macOS, and requires Java 11 or newer.

`transport.tcp.keep_interval`::
Configures the `TCP_KEEPINTVL` option for this socket,
which determines the time in seconds between sending TCP keepalive probes.
Defaults to `network.tcp.keep_interval` if set, or the system default otherwise.

> May not be greater than 300. In case where the system default is higher than 300,
> the value is automatically lowered to 300. Only applicable on Linux and macOS,

Reviewer 2: ```suggestion
This value cannot exceed `300` seconds. In cases where the system default is higher than `300`,
the value is automatically lowered to `300`. Only applicable on Linux and macOS,
```

`transport.compress`::
Set to `true` to enable compression (`DEFLATE`) between
all nodes. Defaults to `false`.

`transport.ping_schedule`::
Schedule a regular application-level ping message
to ensure that transport connections between nodes are kept alive. Defaults to
`5s` in the transport client and `-1` (disabled) elsewhere. It is preferable
to correctly configure TCP keep-alives instead of using this feature, because
TCP keep-alives apply to all kinds of long-lived connections and not just to
transport connections.

It also uses the common
<<modules-network,network settings>>.

`transport.tcp.no_delay`::
Enable or disable the https://en.wikipedia.org/wiki/Nagle%27s_algorithm[TCP no delay]
setting. Defaults to `network.tcp.no_delay`.

`transport.tcp.keep_alive`::
Configures the `SO_KEEPALIVE` option for this socket, which
determines whether it sends TCP keepalive probes.
Defaults to `network.tcp.keep_alive`.

`transport.tcp.keep_idle`::
Configures the `TCP_KEEPIDLE` option for this socket, which
determines the time in seconds that a connection must be idle before
starting to send TCP keepalive probes. Defaults to `network.tcp.keep_idle`, which

> is the system default. May not be greater than 300. In case where the system default

Reviewer: This is awkward to explain; my best idea so far is this:

```suggestion
Defaults to `network.tcp.keep_idle` if set, or the system default otherwise.
May not be greater than 300. In case where the system default
```

Maybe the docs folks can think of a better way to explain this?
